src/Workspaces/Remote/ServiceHub/Services/CustomMessageHandler/RemoteCustomMessageHandlerService.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CustomMessageHandler;
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+internal sealed partial class RemoteCustomMessageHandlerService : BrokeredServiceBase, IRemoteCustomMessageHandlerService
+{
+    internal sealed class Factory : FactoryBase<IRemoteCustomMessageHandlerService>
+    {
+        protected override IRemoteCustomMessageHandlerService CreateService(in ServiceConstructionArguments arguments)
+            => new RemoteCustomMessageHandlerService(arguments);
+    }
+
+    public RemoteCustomMessageHandlerService(in ServiceConstructionArguments arguments)
+        : base(arguments)
+    {
+    }
+
+    public ValueTask<string> HandleCustomMessageAsync(
+        Checksum solutionChecksum,
+        string assemblyFolderPath,
+        string assemblyFileName,
+        string typeFullName,
+        string jsonMessage,
+        DocumentId? documentId,
+        CancellationToken cancellationToken)
+    {
+        var service = this.GetWorkspace().Services.GetRequiredService<ICustomMessageHandlerService>();
```

Comment:
@tmat should we be using `Solution.Services` instead of accessing off of the workspace (and all other places where this gets called?

Reply:
Yes, whenever we have access to Solution, Solution.Services is preferred.

Reply:
I am now using the solution wherever we have one. But the load/unload methods don't need a solution, so I am still going through the workspace for those